volatility3/framework/plugins/windows/cachedump.py

```diff
@@ -1,187 +1,23 @@
-        pad = 2 * ((uname_len / 2) % 2)
-        domain_offset = int(uname_offset + uname_len + pad)
-        pad = 2 * ((domain_len / 2) % 2)
-        domain_name_offset = int(domain_offset + domain_len + pad)
-        hashh = dec_data[:0x10]
-        username = dec_data[uname_offset : uname_offset + uname_len].decode(
-            "utf-16-le", "replace"
-        )
-        domain = dec_data[domain_offset : domain_offset + domain_len].decode(
-            "utf-16-le", "replace"
-        )
-        domain_name = dec_data[
-            domain_name_offset : domain_name_offset + domain_name_len
-        ].decode("utf-16-le", "replace")
-
-        return (username, domain, domain_name, hashh)
-
-    def _generator(self, syshive, sechive):
-        if not syshive or not sechive:
-            if syshive is None:
-                vollog.warning("Unable to locate SYSTEM hive")
-            if sechive is None:
-                vollog.warning("Unable to locate SECURITY hive")
-            return None
-
-        bootkey = hashdump.Hashdump.get_bootkey(syshive)
-        if not bootkey:
-            vollog.warning("Unable to find bootkey")
-            return None
-
-        kernel = self.context.modules[self.config["kernel"]]
-
-        vista_or_later = versions.is_vista_or_later(
-            context=self.context, symbol_table=kernel.symbol_table_name
-        )
-
-        lsakey = lsadump.Lsadump.get_lsa_key(sechive, bootkey, vista_or_later)
-        if not lsakey:
-            vollog.warning("Unable to find lsa key")
-            return None
-
-        nlkm = self.get_nlkm(sechive, lsakey, vista_or_later)
-        if not nlkm:
-            vollog.warning("Unable to find nlkma key")
-            return None
-
-        cache = hashdump.Hashdump.get_hive_key(sechive, "Cache")
-        if not cache:
-            vollog.warning("Unable to find cache key")
-            return None
-
-        for cache_item in cache.get_values():
-            if cache_item.Name == "NL$Control":
-                continue
-
-            try:
-                data = sechive.read(cache_item.Data + 4, cache_item.DataLength)
-            except exceptions.InvalidAddressException:
-                continue
-
-            if not data:
-                continue
-
-            (
-                uname_len,
-                domain_len,
-                domain_name_len,
-                enc_data,
-                ch,
-            ) = self.parse_cache_entry(data)
-            # Skip if nothing in this cache entry
-            if uname_len == 0 or len(ch) == 0:
-                continue
-            dec_data = self.decrypt_hash(enc_data, nlkm, ch, not vista_or_later)
-
-            (username, domain, domain_name, hashh) = self.parse_decrypted_cache(
-                dec_data, uname_len, domain_len, domain_name_len
+    def __getattr__(self, *args, **kwargs):
+        warnings.warn(
+            FutureWarning(
+                "This plugin is now called windows.registry.cachedump.Cachedump"
```

Comment:
Might also be worth letting people know this will go away (so they have a reason to change) and including a deprecation date (six months plus a week seems to be what @atcuno 's been going with, so happy to standardize on that).

Reply:
Yeah, I mentioned in another comment that I figured out if I changed it to `__getattribute__`, it printed on the command line if you called the deprecated version directly (`vol.py ... windows.hashdump`), but if I modify cachedump to import windows.hashdump instead of windows.registry.hashdump (to simulate somone writing code that imports the deprecated file), then when it calls the hashdump.get_bootkey() classmethod, no warning is printed, which is why you had originally suggested using getattr.

I pushed up a version with getattribute and the warning date. Not sure if you have any ideas on getting the message if the old code is imported. If not then I guess this is ready.

Reply:
Hmmmm, that's a little strange?  I guess you could implement both `getattribute` and `getattr`, but I dunno why it would need to be looking that up (and I thought that would pass through `getattribute` anyway?)

Reply:
I guess the classmethod decorator falls back to the class it's actually defined in?  I'd still expect the lookup to get called though?

Reply:
See my comment below for something that works (although, hard to tell how hacky it is, probably needs checking on a bunch of versions of python)...